packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/fill-row-data-indent-calculations.ts

```diff
@@ -0,0 +1,115 @@
+         * I have no child AND
+         * previous row's indent is not smaller then mine,
+         * then draw "up" line
+         */
+        if (prevRow) {
+          const isPrevIndentNotSmallerThanMe = prevRow.indent >= indent;
+
+          hasUp = Boolean(!iHaveChild && isPrevIndentNotSmallerThanMe);
+        }
+      }
+      if (hasUp) {
+        arr[i] = "up";
+      }
+    }
+
+    property.verticalLinesForEachIndent = arr;
+  }
+};
+
+/**
+ * this **SHOULD** be runned after `markUpperHalvesOfIndentationLines`
+ * because it uses the previously calculated/marked "up" halves of indentation lines
+ * to calculate the down halves
+ */
+const markBottomHalvesOfIndentationLines = (rowData: PropertyRow[]) => {
+  /**
+   * For each row, starting from the last row and going backwards.
+   * Previous function called `markUpperHalvesOfIndentationLines` already calculated & marked all "up" parts of the lines.
+   * So instead of a complex logic to calculate "down" parts,
+   * for each row, we look at the row below to see if there are lines that needs to continue upwards.
+   * Reason for going backwars is, each row needs to check if the row below has a line which should continue upwards.
+   * So we start from bottom, and keep drawing each line until they end
+   */
+  for (let row = rowData.length - 1; row >= 0; row--) {
+    const property = rowData[row] as PropertyRow;
+
+    const { verticalLinesForEachIndent } = property;
+
+    const { indent } = property;
+
+    const nextRow = rowData[row + 1];
+    const prevRow = rowData[row - 1];
+
+    /**
+     * If there is no nextRow (means this is the last row),
+     * or newRow has no depth, no need to check if we should have "down" half of the line for this row.
+     * Because it's impossible in this case
+     */
+    if (nextRow && nextRow.depth > 0) {
+      // for each indentation level
+      for (let i = 0; i <= indent; i++) {
+        let hasDown = false;
+
+        // if there are lines that needs to continue upwards at the row below
+        if (
+          nextRow.verticalLinesForEachIndent[i] === "up" ||
+          nextRow.verticalLinesForEachIndent[i] === "full"
+        ) {
+          hasDown = true;
+        }
+
+        let shouldThisBeAFullLine = false;
+        const isDrawingCurrentIndent = i === indent;
+
+        if (!isDrawingCurrentIndent && prevRow?.depth) {
+          shouldThisBeAFullLine = true;
+        }
+
+        if (hasDown) {
+          const wasThisMarkedAsUp = verticalLinesForEachIndent[i] === "up";
+          const shouldBeFull = wasThisMarkedAsUp || shouldThisBeAFullLine;
+          verticalLinesForEachIndent[i] = shouldBeFull ? "full" : "down";
+        }
+      }
+    }
+  }
+};
+
+/**
+ * Calculates & fills `verticalLinesForEachIndent` for each property row
+ * It uses `indent` and `children` of each property row for these calculations
+ */
+export const fillRowDataIndentCalculations = (rowData: PropertyRow[]) => {
```

Comment:
Addressed here, this will also change with the alternative solution, I don't think it's worth trying to optimize this right now (I'd prefer to switch to the alternative solution if we see any performance issue about this in future) https://github.com/hashintel/hash/pull/1266#discussion_r1011577665

Reply:
I think for me the main thing is readability/maintainability over performance, but sure, I get why you may not want to make this change now. Thanks for adding the comment.